dask_expr/groupby.py

```diff
@@ -0,0 +1,422 @@
+import functools
+
+import numpy as np
+from dask.base import normalize_token
+from dask.dataframe.core import (
+    _concat,
+    is_dataframe_like,
+    is_series_like,
+    make_meta,
+    meta_nonempty,
+    no_default,
+)
+from dask.dataframe.groupby import (
+    _agg_finalize,
+    _apply_chunk,
+    _build_agg_args,
+    _groupby_aggregate,
+    _groupby_apply_funcs,
+    _normalize_spec,
+)
+from dask.utils import M
+
+from dask_expr.collection import DataFrame, new_collection
+from dask_expr.reductions import ApplyConcatApply
+
+###
+### Groupby-aggregation expressions
+###
+
+
+class BaseAggregation(ApplyConcatApply):
+    """Groupby-aggregation base class
+
+    This class does very little besides define some
+    common properties used by `SingleAggregation` and
+    `GroupbyAggregation`.
+
+    See Also
+    --------
+    SingleAggregation
+    GroupbyAggregation
+    """
+
+    @property
+    def dropna(self) -> dict:
+        dropna = self.operand("dropna")
+        if dropna is not None:
+            return {"dropna": dropna}
+        return {}
+
+    @property
+    def observed(self) -> dict:
+        observed = self.operand("observed")
+        if observed is not None:
+            return {"observed": observed}
+        return {}
+
+    @property
+    def levels(self):
+        if isinstance(self.by, (tuple, list)) and len(self.by) > 1:
+            levels = list(range(len(self.by)))
+        else:
+            levels = 0
+        return levels
```

Comment:
I'm not sure I understand what is being highlighted here? We don't need the `levels` property, but we do need to convert the `by` operand to the corresponding `levels` in a few places.

Reply:
Ah sorry, let me focus my comment on the two methods just above `dropna` and `observed` and skip `levels` for now.  It looks like we're storing something like a boolean, but we access it as a dictionary.  

My perference would be that we not do the bool->dict conversion in the same name as the parameter.  Right now I mostly assume that `expr.foo` is the same thing as `expr.operands("foo")`.  We're breaking this assumption here.  I'm ok being non-rigid, but in this case I don't see a very compelling case.  It might make as much sense to do this dict expansion where we use `self.dropna` or, if we really wanted to, maybe we make a `dropna_dict` method or something with a name that is different than the parameter.

Reply:
I'm not sure if that was any more clear.  Let me know if now.

Reply:
Ah I see. Yeah, that was mostly just a copy from dask.dataframe. I’m also not a fan of that pattern, and will be happy to clean it up.